File: python_plots_generator/Server/DTOS/PieDTO.py
from flask_restx import fields, Namespace

from Config.constants import FORMATTERS_LIST

"""

    {
        "src": {
            "path": "spark/temp/evol/cadiz/evol",
            "names": {
                "lower_bound": "min_value",
                "upper_bound": "max_value",
                "value": "count"
            }
        },
        "dest" : {
            "path": "temp/evol/cadiz/evol",
            "filename": "plot_test",
            "export_png": true
        },
        "style": {
            "lettering": {
                "title": "TITLE TEST",
                "subtitle": "SUBTITLE TEST"
            },
            "margin": {
                "left": 120,
                "right": 120,
                "top": 100,
                "bottom": 100
            },
            "show_legend": true
        }
    }


"""

class PieDTO:
    def __init__(self, ns: Namespace):
        self.post_input = _create_input_post_dto(ns)


def _create_input_post_dto(ns: Namespace):
    return ns.model('PieInput', {
        'src': fields.Nested(ns.model('PieSrc', {
            'path': fields.String(required=True, description="Relative route to data"),
            'names': fields.Nested(ns.model('PieSrcNames', {
                'lower_bound': fields.String(reqired=True, description="Set lower bound column name"),
                'upper_bound': fields.String(reqired=True, description="Set upper bound column name"),
                'value': fields.String(reqired=True, description="Set value column name"),
            }), required=True)
        })),
        'dest': fields.Nested(ns.model('PieDest', {
            'path': fields.String(required=True, description="Relative output route"),
            'filename': fields.String(required=True, description="Filename of the chart"),
            'export_png': fields.Boolean(required=True, description="Export png"),
        }), required=True),
        'style': fields.Nested(ns.model('PieStyle', {
            'lettering': fields.Nested(ns.model('PieStyleLettering', {
                'title': fields.String(required=True, description="Title of the chart"),
                'subtitle': fields.String(required=False, description="Subtitle of the chart")
            }), required=True),
            'margin': fields.Nested(ns.model('PieStyleMargin', {
                'left': fields.Float(required=True, description="Margin left"),
                'right': fields.Float(required=True, description="Margin right"),
                'top': fields.Float(required=True, description="Margin top"),
                'bottom': fields.Float(required=True, description="Margin bottom")
            }), required=False),
            'show_legend': fields.Boolean(required=True, description="Show legend"),
        }), required=True)
    })

File: big-data-core/src/main/scala/Config/PlotGenerationConf/Execution/DTO/PieDTO.scala
package Config.PlotGenerationConf.Execution.DTO

import upickle.default.{ReadWriter, macroRW}
import upickle.implicits.key


case class PieDTOSrcNamesComponent(
  @key("lower_bound") lowerBound: String,
  @key("upper_bound") upperBound: String,
  @key("value") value: String
)
object PieDTOSrcNamesComponent { implicit val rw: ReadWriter[PieDTOSrcNamesComponent] = macroRW }

case class PieDTOSrc(
  @key("path") path: String,
  @key("names") names: PieDTOSrcNamesComponent
)
object PieDTOSrc { implicit val rw: ReadWriter[PieDTOSrc] = macroRW }

case class PieDTODest(
  @key("path") path: String,
  @key("filename") filename: String,
  @key("export_png") exportPng: Boolean
)
object PieDTODest { implicit val rw: ReadWriter[PieDTODest] = macroRW }

case class PieDTOLettering(
  @key("title") title: String,
  @key("subtitle") subtitle: String
)
object PieDTOLettering { implicit val rw: ReadWriter[PieDTOLettering] = macroRW }

case class PieDTOMargin(
  @key("left") left: Float,
  @key("right") right: Float,
  @key("top") top: Float,
  @key("bottom") bottom: Float
)
object PieDTOMargin { implicit val rw: ReadWriter[PieDTOMargin] = macroRW }

case class PieDTOStyle(
  @key("lettering") lettering: PieDTOLettering,
  @key("margin") margin: Option[PieDTOMargin],
  @key("show_legend") showLegend: Boolean
)
object PieDTOStyle { implicit val rw: ReadWriter[PieDTOStyle] = macroRW }

case class PieDTO (
  @key("src") src: PieDTOSrc, 
  @key("dest") dest: PieDTODest, 
  @key("style") style: PieDTOStyle
)
object PieDTO { implicit val rw: ReadWriter[PieDTO] = macroRW }
